### PR TITLE
switch to triton ascend 3.2.1

### PR DESCRIPTION
**Pull request by kevin-hongkai**

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -19,7 +19,6 @@ dependencies = [
 
 [project.optional-dependencies]
 npu = [
-    "byted-triton-x >= 3.2.0.post24"
 ]
```
[the remark this answers is not in the document]
Comment (kevin-hongkai) [reply pyproject.toml:23]: 已经添加了byted-triton-x>= 3.2.1

File: mojo_opset/backends/ttx/functions/normalization.py [hunk 2]
```diff
@@ -23,7 +23,7 @@ def forward(
     ) -> torch.Tensor:
         # FIXME: Currently, MojoNormFunction base class does not define fields like 'offset', so they are hardcoded here temporarily.
         offset = 0.0
-        casting_mode = "llama"
+        casting_mode = "gemma"
```
[the remark this answers is not in the document]
Comment (kevin-hongkai) [reply mojo_opset/backends/ttx/functions/normalization.py]: rms_norm 算子实现内有多个 cast_mode，默认走的 llama_cast_mode 存在形如 norm([x1.to](http://x1.to/)(fp32)).to(x1.dtype) 的操作； 作为标杆的 torch 自带的 F.rms_norm 则是使用 fp32 高精度计算下来的，这会导致散点式的精度误差。当前可以将 cast_mode 配置为 gemma_mode 以规避该问题。

File: mojo_opset/backends/ttx/kernels/npu/quant.py [hunk 3]
```diff
@@ -141,6 +141,6 @@ def scale_dynamic_quant_kernel(
             scaled_vals = input_vals * scale_vals
             quant_vals = scaled_vals / current_quant_scale[:, None]
             quant_vals = tl.where(quant_vals < 0, quant_vals - 0.5, quant_vals + 0.5)
-            quant_vals_int8 = tl.cast(quant_vals, dtype=tl.int8, overflow_mode="saturate")
+            quant_vals_int8 = tl.cast(quant_vals, dtype=tl.int8)
```
[the remark this answers is not in the document]
Comment (kevin-hongkai) [reply mojo_opset/backends/ttx/kernels/npu/quant.py:154]: 不再需要添加overflow_mode="saturate"，triton-ascend新版本也不再支持此字段的传入

File: mojo_opset/backends/ttx/kernels/npu/sdpa.py [hunk 4]
```diff
@@ -65,8 +65,8 @@ def _sdpa_infer_inner(
         # qk += (1 - mask.to(tl.float32)) * (-1e6)
         # qk = tl.where(mask, qk, float("-inf"))
         qk = tl.where(mask, qk, -1e6)
-
-        m_ij = tl.maximum(m_i, tl.max(qk, 1))  # Scaled max
+        m_ij = tl.maximum(m_i, tl.max(qk, 1, propagate_nan=tl.PropagateNan.ALL), propagate_nan=tl.PropagateNan.ALL)
+        #m_ij = tl.maximum(m_i, tl.max(qk, 1))  # Scaled max
```
[the remark this answers is not in the document]
Comment (kevin-hongkai) [reply mojo_opset/backends/ttx/kernels/npu/sdpa.py:69]: 增加propagate_nan=tl.PropagateNan.ALL：底层的指令max针对nan的处理是 max(nan, a) = nan 或者 max(nan, nan)=nan,也就是 arith.maximumf: 传播 NaN：只要任一操作数是 NaN，结果就是 NaN, maximumf(1.0, NaN) = NaN的语义
因此前端通过propagate nan来区分，propagate nan=propagate.ALL->默认生成arith.maximumf, 否则是arith.maxnumf， 因为底层指令不支持，所以会插入一些模拟指令处理边界

File: mojo_opset/backends/ttx/kernels/npu/over_encoding/n_gram.py [hunk 5]
```diff
@@ -226,7 +226,7 @@ def n_gram_decode_kernel(
                     output_ids + bid * output_stride_0 + tl.arange(0, MTP_STEP)[:, None] * output_stride_1 + block_offsets[None, :] * output_stride_2,
                     # output_ids + bid * output_stride_0 + output_offsets,
                     n_gram_ids.to(output_ids.dtype.element_ty),
-                    mask=tl.view(block_mask, (1, BLOCK_SIZE_N)),
+                    mask= block_mask_1  #tl.view(block_mask, (1, BLOCK_SIZE_N)),
```
[the remark this answers is not in the document]
Comment (kevin-hongkai) [reply mojo_opset/backends/ttx/kernels/npu/over_encoding/n_gram.py]: for resolve random precision error. OK，delete the comment

**Comment by kevin-hongkai** [comment]
> ### Must fix
> * **[BLOCKER] RMSNorm default casting_mode changed from "llama" to "gemma"** -- `mojo_opset/backends/ttx/functions/normalization.py:26` -- This silently changes numerical behavior of RMSNorm for all callers (gemma casts weight differently). Either revert or expose the field on the base class instead of flipping the hardcoded default.
-----rms_norm 算子实现内有多个 cast_mode，默认走的 llama_cast_mode 存在形如 norm([x1.to](http://x1.to/)(fp32)).to(x1.dtype) 的操作； 作为标杆的 torch 自带的 F.rms_norm 则是使用 fp32 高精度计算下来的，这会导致散点式的精度误差。当前可以将 cast_mode 配置为 gemma_mode 以规避该问题
> * **[BLOCKER] int8 quant cast lost `overflow_mode="saturate"`** -- `mojo_opset/backends/ttx/kernels/npu/quant.py:154` -- Without saturation, out-of-range values wrap; the manual `+/- 0.5` rounding can still exceed [-127,127] (e.g. NaNs/inf or scale=0 paths), producing wrong int8 values. Restore saturate or guarantee saturation upstream.
------不再需要添加overflow_mode="saturate"，triton-ascend新版本也不再支持此字段的传入
> * **[BLOCKER] Perf test dropped reference comparison** -- `mojo_opset/tests/perf/test_indexer.py:32` -- Removing the torch reference call makes the perf test purely one-sided; intentional? If yes, rename/document; otherwise restore `indexer_ref` perf so regressions vs reference remain detectable.
------torch在perf时容易卡死，耗时很常，没必要做torch性能
> * **[BLOCKER] `compile_hint` string changed: `"tile_cube_loop"` -> `"hivm.tile_mix_cube_num"`** -- `mojo_opset/backends/ttx/kernels/npu/sdpa.py:85` -- All other call sites kept their original hint strings (e.g. `"dot_pad_only_k"`); only this one was renamed. Confirm this is the intended new hint name and not a copy-paste; a wrong hint silently degrades performance on a hot decode path.

> * **[BLOCKER] `MojoIndexer.quant` call signature changed and result reshaped** -- `mojo_opset/experimental/operators/indexer.py:92-94` -- Switched from `quant(q, None)` to `quant(q)` plus a new `q_scale.squeeze(-1)`; if `quant` still returns a 3D scale this squeeze is a no-op or wrong depending on shape. Add a shape assertion or align with the `k_scale.dim()==3` branch handling.
------经测试，indexer中接口变更了，但是上层调用还是使用quant(q, None)，会报错参数不匹配
> 
> ### Suggestions
> Suggestions (5)
> * **[MAJOR] Stray `device_print` removal but no test for the previous output** -- `mojo_opset/backends/ttx/kernels/npu/over_encoding/n_gram.py:188` -- Good removal of debug residue; verify there is a test exercising MTP_STEP>1 path that previously relied on the printed value being incidental.
-------原来为了解决load在IR消失的问题，才加device_print；新版本triton-ascend已解决，可去除device_print
> * **[MAJOR] `sync_solver=False` added to grouped matmul launches** -- `mojo_opset/backends/ttx/kernels/npu/group_gemm.py:227,327` -- This changes synchronization semantics; please document why it is safe (no cross-core dependency) and add a comment in the kernel.
-----triton-ascend中groupgemm性能退化20%，是由于同步引起的，cube类不需要加同步，可使用sync_solver编译选项先去除同步； 后面版本的编译器已识别该问题，后续版本会解决性能退化问题。
> * **[MAJOR] `enable_ubuf_saving=True` added unconditionally** -- `mojo_opset/backends/ttx/kernels/npu/swa.py:633,1027` -- Confirm this does not regress shapes that previously fit in UB; consider gating on tile size.
-----需要添加enable_ubuf_saving，否则会ub overflow

> * **[MINOR] `int64 -> int32` change in test util** -- `mojo_opset/tests/perf/test_linear.py:22` -- Silent dtype change; ensure the kernel under test accepts int32 group sizes (likely yes, but worth a comment).
----测试发现，perf的用例test_linear.py 中int64类型会报错类型不匹配，改为int32修复
> 
> ### Notes
> * [CHECK] All `tl.extra.cann.extension.*` and `tl.extra.cann.math.tanh` symbols exist in the pinned `byted-triton-x>=3.2.1`; if not, kernels will fail at compile time only when first invoked.
------triton-ascend新版本中为兼容triton社区，NPU的所有扩展接口都改为tl.extra.cann.extention路径
> * [CHECK] The `propagate_nan=tl.PropagateNan.ALL` change for `tl.max`/`tl.maximum` may alter softmax behavior when inputs contain NaN (previously NaNs may have been silently dropped); verify this matches intended semantics on attention masks set to `-inf`.
------max添加propagate_nan=tl.PropagateNan.ALL可增加性能。编译器解释原因为：增加propagate_nan=tl.PropagateNan.ALL：底层的指令max针对nan的处理是 max(nan, a) = nan 或者 max(nan, nan)=nan,也就是 arith.maximumf: 传播 NaN：只要任一操作数是 NaN，结果就是 NaN, maximumf(1.0, NaN) = NaN的语义
因此前端通过propagate nan来区分，propagate nan=propagate.ALL->默认生成arith.maximumf, 否则是arith.maxnumf， 因为底层指令不支持，所以会插入一些模拟指令处理边界

File: .github/workflows/ascend_accuracy_ci.yml [hunk 6]
```diff
@@ -56,6 +56,9 @@ jobs:
           pip install -e .\[npu\] --index-url https://bytedpypi.byted.org/simple
           python3 -m build .
           pip install dist/*.whl --force-reinstall --no-deps
+          pip uninstall -y byted-triton-x triton
```
Comment (zhangjihang-BD) [inline .github/workflows/ascend_accuracy_ci.yml]: 这个应该不需要手动卸载安装？应该通过更新pyproject.toml的方式。

File: pyproject.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (zhangjihang-BD) [reply pyproject.toml:23]: 缩进稍微对齐下，另外是否考虑下午说的3.2.2版本

File: .github/workflows/ascend_accuracy_ci.yml (hunk 6 above)
Comment (kevin-hongkai) [reply .github/workflows/ascend_accuracy_ci.yml]: 已去除

File: pyproject.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (kevin-hongkai) [reply pyproject.toml:23]: 已缩进对齐，建议保留3.2.1版本，如果3.2.2版本发布或有更高版本>=会自动寻找最高版本

**Review by zhangjihang-BD — APPROVED** [review]
